lib/spells/effects/Timed.cpp

```diff
@@ -226,8 +226,8 @@ void Timed::convertBonus(const Mechanics * m, int32_t & duration, std::vector<Bo
 		if((nb.sid == SpellID::SHIELD || nb.sid == SpellID::AIR_SHIELD) && (nb.type == Bonus::GENERAL_DAMAGE_REDUCTION))
 			nb.val = 100 - nb.val;
 		//we need to know who cast Bind
-		else if(nb.sid == SpellID::BIND && nb.type == Bonus::BIND_EFFECT && m->caster->getHeroCaster() == nullptr)
-			nb.additionalInfo = m->caster->getCasterUnitId();
+		else if(nb.sid == SpellID::BIND && nb.type == Bonus::BIND_EFFECT && m->caster->getBattleCasterID() >= 0)
```

Comment:
Is hero both adventure and battle caster? I would answer yes. Hero can cast during battle. But it is not a unit.
Detecting that caster is unit just because m->caster->getHeroCaster() == nullptr is also wrong. The one who is not a hero may still be not a unit. For now artifacts can cast so probably it is made in a way that the hero actually cast these spells. Moat can cast, ground can theoretically cast something in future. Obstacles? Binding trap. You step into it and it binds you for a few turns. So some getCasterType() which return HERO, UNIT, OBSTACLE and so on. And if you checked that caster is UNIT you can ask getCasterUnitID. Quite logical as for me

Reply:
If we use getCasterType we actually can use getCasterID as we can understand what type this ID is by caster type. getCasterHero is also possible if we want to know whether caster is a hero. Kind of dynamic cast as I said before.

Reply:
Your controversal commit is moved outside of this PR.